File: src/main/java/com/joecollins/graphics/components/lowerthird/SummaryWithLabels.java
package com.joecollins.graphics.components.lowerthird;

import com.joecollins.bindings.Binding;
import com.joecollins.bindings.IndexedBinding;
import com.joecollins.graphics.utils.StandardFont;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.GridLayout;
import java.awt.LayoutManager;
import java.util.ArrayList;
import java.util.List;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;
import javax.swing.border.MatteBorder;

public class SummaryWithLabels extends JPanel {

  private Binding<Integer> numEntriesBinding = () -> 1;
  private IndexedBinding<Entry> entriesBinding =
      IndexedBinding.singletonBinding(new Entry(Color.WHITE, "NO RESULT YET", "WAITING..."));

  private List<EntryPanel> entryPanels = new ArrayList<>();

  public SummaryWithLabels() {
    setBackground(Color.WHITE);
    setLayout(new SummaryLayout());
    setBorder(new MatteBorder(1, 1, 1, 1, Color.BLACK));
    setNumEntriesBinding(numEntriesBinding);
    setEntriesBinding(entriesBinding);
  }

  public static class Entry {
    private final Color color;
    private final String label;
    private final String value;

    public Entry(Color color, String label, String value) {
      this.color = color;
      this.label = label;
      this.value = value;
    }
  }

  public void setNumEntriesBinding(Binding<Integer> numEntriesBinding) {
    this.numEntriesBinding.unbind();
    this.numEntriesBinding = numEntriesBinding;
    this.numEntriesBinding.bindLegacy(
        size -> {
          while (entryPanels.size() < size) {
            EntryPanel newPanel = new EntryPanel();
            add(newPanel);
            entryPanels.add(newPanel);
          }
          while (entryPanels.size() > size) {
            remove(entryPanels.remove(size.intValue()));
          }
        });
  }

  int getNumEntries() {
    return this.entryPanels.size();
  }

  public void setEntriesBinding(IndexedBinding<Entry> entriesBinding) {
    this.entriesBinding.unbind();
    this.entriesBinding = entriesBinding;
    this.entriesBinding.bind(
        (idx, entry) -> {
          entryPanels
              .get(idx)
              .topLabel
              .setForeground(entry.color.equals(Color.BLACK) ? Color.WHITE : entry.color);
          entryPanels.get(idx).bottomPanel.setBackground(entry.color);
          entryPanels
              .get(idx)
              .bottomLabel
              .setForeground(entry.color.equals(Color.WHITE) ? Color.BLACK : Color.WHITE);
          entryPanels.get(idx).topLabel.setText(entry.label);
          entryPanels.get(idx).bottomLabel.setText(entry.value);
        });
  }

  Color getEntryColor(int index) {
    return entryPanels.get(index).bottomPanel.getBackground();
  }

  String getEntryLabel(int index) {
    return entryPanels.get(index).topLabel.getText();
  }

  String getEntryValue(int index) {
    return entryPanels.get(index).bottomLabel.getText();
  }

  private class EntryPanel extends JPanel {
    private final JPanel topPanel =
        new JPanel() {
          {
            setBackground(Color.BLACK);
            setLayout(new GridLayout(1, 1));
          }
        };
    private final JLabel topLabel =
        new JLabel() {
          {
            setFont(StandardFont.readNormalFont(16));
            setHorizontalAlignment(JLabel.CENTER);
            setForeground(Color.WHITE);
            setBorder(new EmptyBorder(3, 0, -3, 0));
          }
        };
    private final JPanel bottomPanel =
        new JPanel() {
          {
            setBackground(Color.WHITE);
            setLayout(new GridLayout(1, 1));
          }
        };
    private final JLabel bottomLabel =
        new JLabel() {
          {
            setFont(StandardFont.readBoldFont(24));
            setHorizontalAlignment(JLabel.CENTER);
            setForeground(Color.BLACK);
            setBorder(new EmptyBorder(4, 0, -4, 0));
          }
        };

    public EntryPanel() {
      add(topPanel);
      add(bottomPanel);
      topPanel.add(topLabel);
      bottomPanel.add(bottomLabel);
      setLayout(new EntryLayout());
    }

    private class EntryLayout implements LayoutManager {

      @Override
      public void addLayoutComponent(String name, Component comp) {}

      @Override
      public void removeLayoutComponent(Component comp) {}

      @Override
      public Dimension preferredLayoutSize(Container parent) {
        return new Dimension(50, 50);
      }

      @Override
      public Dimension minimumLayoutSize(Container parent) {
        return new Dimension(10, 50);
      }

      @Override
      public void layoutContainer(Container parent) {
        int width = parent.getWidth();
        int height = parent.getHeight();
        int mid = height * 2 / 5;
        topPanel.setLocation(0, 0);
        topPanel.setSize(width, mid);
        bottomPanel.setLocation(0, mid);
        bottomPanel.setSize(width, height - mid);
      }
    }
  }

  private class SummaryLayout implements LayoutManager {

    @Override
    public void addLayoutComponent(String name, Component comp) {}

    @Override
    public void removeLayoutComponent(Component comp) {}

    @Override
    public Dimension preferredLayoutSize(Container parent) {
      return new Dimension(512, 50);
    }

    @Override
    public Dimension minimumLayoutSize(Container parent) {
      return new Dimension(50, 50);
    }

    @Override
    public void layoutContainer(Container parent) {
      int width = parent.getWidth() - 2;
      int height = parent.getHeight() - 2;
      for (int i = 0; i < entryPanels.size(); i++) {
        int left = width * i / entryPanels.size();
        int right = width * (i + 1) / entryPanels.size();
        entryPanels.get(i).setLocation(left + 1, 1);
        entryPanels.get(i).setSize(right - left, height);
      }
    }
  }
}

File: src/main/java/com/joecollins/graphics/components/lowerthird/SummaryWithLabels.kt
package com.joecollins.graphics.components.lowerthird

import com.joecollins.bindings.Binding
import com.joecollins.graphics.utils.StandardFont
import java.awt.Color
import java.awt.Component
import java.awt.Container
import java.awt.Dimension
import java.awt.GridLayout
import java.awt.LayoutManager
import java.util.ArrayList
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.border.EmptyBorder
import javax.swing.border.MatteBorder

class SummaryWithLabels(
    private val entriesBinding: Binding<List<Entry>>
) : JPanel() {
    private val entryPanels: MutableList<EntryPanel> = ArrayList()

    class Entry(val color: Color, val label: String, val value: String)

    internal val numEntries: Int
        get() = entryPanels.size

    internal fun getEntryColor(index: Int): Color {
        return entryPanels[index].bottomPanel.background
    }

    internal fun getEntryLabel(index: Int): String {
        return entryPanels[index].topLabel.text
    }

    internal fun getEntryValue(index: Int): String {
        return entryPanels[index].bottomLabel.text
    }

    private inner class EntryPanel : JPanel() {
        private val topPanel: JPanel = object : JPanel() {
            init {
                background = Color.BLACK
                layout = GridLayout(1, 1)
            }
        }
        val topLabel: JLabel = object : JLabel() {
            init {
                font = StandardFont.readNormalFont(16)
                horizontalAlignment = CENTER
                foreground = Color.WHITE
                border = EmptyBorder(3, 0, -3, 0)
            }
        }
        val bottomPanel: JPanel = object : JPanel() {
            init {
                background = Color.WHITE
                layout = GridLayout(1, 1)
            }
        }
        val bottomLabel: JLabel = object : JLabel() {
            init {
                font = StandardFont.readBoldFont(24)
                horizontalAlignment = CENTER
                foreground = Color.BLACK
                border = EmptyBorder(4, 0, -4, 0)
            }
        }

        private inner class EntryLayout : LayoutManager {
            override fun addLayoutComponent(name: String, comp: Component) {}
            override fun removeLayoutComponent(comp: Component) {}
            override fun preferredLayoutSize(parent: Container): Dimension {
                return Dimension(50, 50)
            }

            override fun minimumLayoutSize(parent: Container): Dimension {
                return Dimension(10, 50)
            }

            override fun layoutContainer(parent: Container) {
                val width = parent.width
                val height = parent.height
                val mid = height * 2 / 5
                topPanel.setLocation(0, 0)
                topPanel.setSize(width, mid)
                bottomPanel.setLocation(0, mid)
                bottomPanel.setSize(width, height - mid)
            }
        }

        init {
            add(topPanel)
            add(bottomPanel)
            topPanel.add(topLabel)
            bottomPanel.add(bottomLabel)
            layout = EntryLayout()
        }
    }

    private inner class SummaryLayout : LayoutManager {
        override fun addLayoutComponent(name: String, comp: Component) {}
        override fun removeLayoutComponent(comp: Component) {}
        override fun preferredLayoutSize(parent: Container): Dimension {
            return Dimension(512, 50)
        }

        override fun minimumLayoutSize(parent: Container): Dimension {
            return Dimension(50, 50)
        }

        override fun layoutContainer(parent: Container) {
            val width = parent.width - 2
            val height = parent.height - 2
            for (i in entryPanels.indices) {
                val left = width * i / entryPanels.size
                val right = width * (i + 1) / entryPanels.size
                entryPanels[i].setLocation(left + 1, 1)
                entryPanels[i].setSize(right - left, height)
            }
        }
    }

    init {
        background = Color.WHITE
        layout = SummaryLayout()
        border = MatteBorder(1, 1, 1, 1, Color.BLACK)

        this.entriesBinding.bind { entries: List<Entry> ->
            while (entryPanels.size < entries.size) {
                val newPanel = EntryPanel()
                add(newPanel)
                entryPanels.add(newPanel)
            }
            while (entryPanels.size > entries.size) {
                remove(entryPanels.removeAt(entries.size))
            }
            entries.onEachIndexed { idx, entry ->
                entryPanels[idx].topLabel.foreground = if (entry.color == Color.BLACK) Color.WHITE else entry.color
                entryPanels[idx].bottomPanel.background = entry.color
                entryPanels[idx].bottomLabel.foreground = if (entry.color == Color.WHITE) Color.BLACK else Color.WHITE
                entryPanels[idx].topLabel.text = entry.label
                entryPanels[idx].bottomLabel.text = entry.value
            }
        }
    }
}
